src/libs/conduit/conduit_node_iterator.hpp

```diff
@@ -54,14 +54,228 @@
 //-----------------------------------------------------------------------------
 // -- conduit includes -- 
 //-----------------------------------------------------------------------------
-#include "conduit_node.hpp"
+#include "conduit_core.hpp"
```

Comment:
is there an issue with order of include?
(could this still be conduit_node, and class not be fowardwad declared below)?

Reply:
`conduit_node.hpp` includes `conduit_node_iterator.hpp` since it returns instances of `NodeIterator` and `NodeConstIterator` by value. I broke the circular dependency by removing `conduit_node.hpp` here. I was about to write that we only needed a forward declaration of `Node` in this file. However, I'm not sure how that's true given that we use `Node&` and `Node*` as template parameters and then write expressions based on those parameters. Maybe this works because anyone working with `conduit_node_iterator.hpp` would also have `conduit_node.hpp`. I did make sure `node_iterator.hpp` is the first include in `node_iterator.cpp`, and that compiles fine (maybe because `conduit_node.hpp` is included before any uses).

After removing `conduit_node.hpp`, I had to add conduit_core.hpp for some other definitions (e.g. `conduit::index_t`).

Thoughts on this?

Do users include the individual `conduit_X.hpp` files, or do they just include `conduit.hpp`?

Reply:
I took a deeper look to better understand.  This is fine b/c both the new iterator classes and the existing ones only use Node refs or Node pointers. 

If I add the following to the base of the new iterator class:

```
    // wont work
    Node WHATISGOINGON()
    {

    }

    // works ok
    Node *WHATISGOINGON_PTR()
    {

    }

```
The first one makes the compiler unhappy, and the second one is fine.
I think we are in good shape, I just wanted to understand what was at play.
Thank!